components/script/dom/datatransferitemlist.rs

```diff
@@ -266,83 +75,78 @@ impl DataTransferItemListMethods for DataTransferItemList {
         data: DOMString,
         mut type_: DOMString,
     ) -> Fallible<Option<DomRoot<DataTransferItem>>> {
-        if let Some(data_store) = self.data_store.root() {
-            // Step 1 If the data store is not in the read/write mode, return null.
-            if !data_store.can_write() {
-                return Ok(None);
-            }
+        // Step 1 If the DataTransferItemList object is not in the read/write mode, return null.
+        let mut option = self.data_store.borrow_mut();
+        let data_store = match option.as_mut() {
+            Some(value) if value.mode() == Mode::ReadWrite => value,
+            _ => return Ok(None),
+        };
 
-            // Step 2.1 If there is already an item in the item list whose kind is text
-            // and whose type string is equal to the method's second argument, throw "NotSupportedError".
-            type_.make_ascii_lowercase();
-            if self
-                .items
-                .borrow()
-                .iter()
-                .any(|item| item.text_type_matches(&type_))
-            {
-                return Err(Error::NotSupported);
-            }
+        // Add an item to the drag data store item list whose kind is text,
+        // whose type string is equal to the value of the method's second argument, converted to ASCII lowercase,
+        // and whose data is the string given by the method's first argument.
+        type_.make_ascii_lowercase();
+        let string = PlainString { data, type_ };
 
-            data_store.invalidate_frozen_types();
+        data_store.add(Kind::Text(string))?;
 
-            // Step 2.2
-            Ok(Some(self.add_item(Kind::Text(data), type_)))
-        } else {
-            Ok(None)
-        }
+        Ok(Some(DataTransferItem::new(&self.global())))
     }
 
     /// <https://html.spec.whatwg.org/multipage/#dom-datatransferitemlist-add>
     fn Add_(&self, data: &File) -> Fallible<Option<DomRoot<DataTransferItem>>> {
-        if let Some(data_store) = self.data_store.root() {
-            // Step 1 If the data store is not in the read/write mode, return null.
-            if !data_store.can_write() {
-                return Ok(None);
-            }
+        // Step 1 If the DataTransferItemList object is not in the read/write mode, return null.
+        let mut option = self.data_store.borrow_mut();
+        let data_store = match option.as_mut() {
+            Some(value) if value.mode() == Mode::ReadWrite => value,
+            _ => return Ok(None),
+        };
+
+        // Add an item to the drag data store item list whose kind is File,
+        // whose type string is the type of the File, converted to ASCII lowercase,
+        // and whose data is the same as the File's data.
+        let mut type_ = data.file_type();
+        type_.make_ascii_lowercase();
+        let binary = Binary {
```

Comment:
I think we actually want to create a BlobImpl here using https://doc.servo.org/script_traits/serializable/struct.BlobImpl.html#method.new_from_file . Then we can store that in the data store and use it to create new File objects as necessary.

Reply:
I suspect the current implementation ends up creating unique blobs each time, and duplicating the contents. This change should avoid that.

Reply:
Yes, it's something that I was looking into, since constructing a File moves the blobimpl, and calling `clone()` ends up with the same outcome of duplicating the contents.

Reply:
Hmm, interesting. I'm ok with leaving this as-is then, and we can revisit at some point in the future.